reproman/interface/tests/test_run.py

```diff
@@ -74,6 +78,87 @@ def test_combine_specs(specs, expected):
     assert _combine_job_specs(specs) == expected
 
 
+@pytest.mark.parametrize(
+    "params,expected",
+    [([], []),
+     (["a=1,2"],
+      [{"a": "1"}, {"a": "2"}]),
+     (["a=1,2", "b=3"],
+      [{"a": "1", "b": "3"},
+       {"a": "2", "b": "3"}]),
+     (["a=1,2", "b=3,4"],
+      [{"a": "1", "b": "3"},
+       {"a": "1", "b": "4"},
+       {"a": "2", "b": "3"},
+       {"a": "2", "b": "4"}]),
+     (["a=1,2=3"],
+      [{"a": "1"},
+       {"a": "2=3"}]),
+     (["a= 1 spaces are preserved   , 2"],
+      [{"a": " 1 spaces are preserved   "},
+       {"a": " 2"}])],
+    ids=["empty", "one", "two, one varying", "two varying", "= in value",
+         "spaces"])
+def test_combine_batch_params(params, expected):
+    actual = list(sorted(_combine_batch_params(params),
+                         key=lambda d: (d.get("a"), d.get("b"))))
+    assert len(actual) == len(expected)
+    assert actual == expected
+
+
+def test_combine_batch_params_glob(tmpdir):
+    tmpdir = str(tmpdir)
+    create_tree(tmpdir, {"aaa": "a",
+                         "subdir": {"b": "b", "c": "c"}})
+    with chpwd(tmpdir):
+        res = sorted(_combine_batch_params(["foo=a*,subdir/*,other"]),
+                     key=lambda d: d["foo"])
+        assert list(res) == [
+            {"foo": "aaa"},
+            {"foo": "other"},
+            {"foo": "subdir/b"},
+            {"foo": "subdir/c"}]
+
+
+def test_combine_batch_params_repeat_key():
+    with pytest.raises(ValueError):
+        list(_combine_batch_params(["a=1", "a=2"]))
+
+
+def test_combine_batch_params_no_equal():
+    with pytest.raises(ValueError):
+        list(_combine_batch_params(["a"]))
+
+
+def test_run_batch_spec_and_params():
+    with pytest.raises(ValueError):
+        run(command="blahbert",
+            batch_spec="anything", batch_parameters="anything")
+
+
+@pytest.mark.parametrize(
+    "params,spec",
+    [([], ""),
+     (["a=1,2"],
+      """\
+- a: '1'
+- a: '2'"""),
+     (["a=1,2", "b=3"],
+      """\
+- a: '1'
+  b: '3'
+- a: '2'
+  b: '3'""")],
+    ids=["empty", "one", "two, one varying"])
+def test_resolve_batch_params_eq(tmpdir, params, spec):
+    fname = op.join(str(tmpdir), "spec.yml")
+    with open(fname, "w") as fh:
+        fh.write(spec)
```

Comment:
Minor FYi: there is create_tree which is used by with_tree decorator

Reply:
Yes, I even use `create_tree` in this PR (ae58628b) :)

I do like `{create,with}_tree`, but I also don't think there's much value in replacing every simple `write` call with it.